### PR TITLE
Add generation of HTML report for different issues

### PR DESCRIPTION
**Pull request by vilchik-elena**

### CODE REVIEW

File: sonar-lits-plugin/src/main/java/com/sonarsource/lits/LitsReport.java [hunk 1]
```diff
@@ -0,0 +1,246 @@
...
+
+      } else {
+        if (componentKey.equals(moduleKey)) {
+          throw new IllegalStateException("Module/project level issues are not supported by LITS plugin");
+        }
+
+        String relativePath = componentKey.substring(moduleKey.length() + 1);
+        this.inputComponent = fileSystem.inputFile(fileSystem.predicates().hasRelativePath(relativePath));
+        if (this.inputComponent == null) {
+          this.inputComponent = fileSystem.inputDir(new File(fileSystem.baseDir(), relativePath));
```
Comment (SonarTech) [inline sonar-lits-plugin/src/main/java/com/sonarsource/lits/LitsReport.java]: MINOR Remove this use of "inputDir"; it is deprecated. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3ACallToDeprecatedMethod)

File: sonar-lits-plugin/src/main/java/com/sonarsource/lits/LitsReport.java [hunk 2]
```diff
@@ -0,0 +1,246 @@
...
+        if (issuesPerLine.containsKey(issue.line)) {
+          issuesPerLine.get(issue.line).add(issue);
+        } else {
+          List<ReportedIssue> perLine = new ArrayList<>();
+          perLine.add(issue);
+          issuesPerLine.put(issue.line, perLine);
+        }
+      }
+
+      retrieveInputComponent(fileSystem, inputComponent, moduleKey);
```
Comment (SonarTech) [inline sonar-lits-plugin/src/main/java/com/sonarsource/lits/LitsReport.java]: MAJOR "NullPointerException" will be thrown when invoking method "retrieveInputComponent()". [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS2259)

File: sonar-lits-plugin/src/main/java/com/sonarsource/lits/LitsReport.java [hunk 3]
```diff
@@ -0,0 +1,246 @@
...
+    boolean isFile;
+
+    private Map<Integer, List<ReportedIssue>> issuesPerLine = new HashMap<>();
+
+    public ResourceReport(String componentKey, Set<ReportedIssue> issues, FileSystem fileSystem) {
+      this.componentKey = componentKey;
+      this.issues = issues;
+
+      // InputFile corresponding to this resource, it's null for new issues
+      InputComponent inputComponent = null;
```
Comment (SonarTech) [inline sonar-lits-plugin/src/main/java/com/sonarsource/lits/LitsReport.java]: MAJOR Rename "inputComponent" which hides the field declared at line 127. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AHiddenFieldCheck)

File: sonar-lits-plugin/src/main/java/com/sonarsource/lits/LitsReport.java [hunk 4]
```diff
@@ -0,0 +1,246 @@
...
+    private void retrieveInputComponent(FileSystem fileSystem, @Nullable InputComponent inputComponent, @Nullable String moduleKey) {
+      if (inputComponent != null) {
+        this.inputComponent = inputComponent;
+
+      } else {
+        if (componentKey.equals(moduleKey)) {
+          throw new IllegalStateException("Module/project level issues are not supported by LITS plugin");
+        }
+
+        String relativePath = componentKey.substring(moduleKey.length() + 1);
```
Comment (SonarTech) [inline sonar-lits-plugin/src/main/java/com/sonarsource/lits/LitsReport.java]: MAJOR A "NullPointerException" could be thrown; "moduleKey" is nullable here. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS2259)

**Comment by SonarTech** [comment]
SonarQube analysis reported 8 issues
* ![MAJOR][MAJOR] 3 major
* ![MINOR][MINOR] 5 minor

Watch the comments in this conversation to review them.

#### 4 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MINOR][MINOR] [DumpPhase.java#L65](https://github.com/SonarSource/sonar-lits/blob/4f50c6c19ca86712946ccf403044fef88a5bf84b/sonar-lits-plugin/src/main/java/com/sonarsource/lits/DumpPhase.java#L65): Remove this use of "InputDir"; it is deprecated. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3ACallToDeprecatedMethod)
1. ![MINOR][MINOR] [DumpPhase.java#L68](https://github.com/SonarSource/sonar-lits/blob/4f50c6c19ca86712946ccf403044fef88a5bf84b/sonar-lits-plugin/src/main/java/com/sonarsource/lits/DumpPhase.java#L68): Remove this use of "InputDir"; it is deprecated. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3ACallToDeprecatedMethod)
1. ![MINOR][MINOR] [DumpPhase.java#L68](https://github.com/SonarSource/sonar-lits/blob/4f50c6c19ca86712946ccf403044fef88a5bf84b/sonar-lits-plugin/src/main/java/com/sonarsource/lits/DumpPhase.java#L68): Remove this use of "inputDir"; it is deprecated. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3ACallToDeprecatedMethod)
1. ![MINOR][MINOR] [DumpPhase.java#L68](https://github.com/SonarSource/sonar-lits/blob/4f50c6c19ca86712946ccf403044fef88a5bf84b/sonar-lits-plugin/src/main/java/com/sonarsource/lits/DumpPhase.java#L68): Remove this use of "file"; it is deprecated. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3ACallToDeprecatedMethod)

[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

File: sonar-lits-plugin/src/main/java/com/sonarsource/lits/IssuesChecker.java [hunk 5]
```diff
@@ -148,14 +152,16 @@ private static void forceDelete(File file) {
...
-  void save() {
+  void save(FileSystem fileSystem) {
     forceDelete(newDumpFile);
     List<String> messages = new ArrayList<>();
     MessageException exception = null;
     if (different) {
       LOG.info("Saving " + newDumpFile);
       Dump.save(dump, newDumpFile);
       messages.add("Issues differences: " + differences);
+      new HtmlReport(fileSystem).print(new LitsReport(differentIssues, fileSystem));
```
Comment (Godin) [inline sonar-lits-plugin/src/main/java/com/sonarsource/lits/IssuesChecker.java]: There are costs to generate report, so maybe it should be optional for the users of this plugin who rely on SonarQube UI and do not need this report.

**Comment by vilchik-elena** [comment]
I close PR as there are many questions about idea and implementation, while it's been about a year developers don't have html report and it's ok

**Pull request closed without merging**